pkg/registry/flowcontrol/rest/storage_flowcontrol.go

```diff
@@ -71,3 +85,80 @@ func (p RESTStorageProvider) v1alpha1Storage(apiResourceConfigSource serverstora
 func (p RESTStorageProvider) GroupName() string {
 	return flowcontrol.GroupName
 }
+
+func (p RESTStorageProvider) PostStartHook() (string, genericapiserver.PostStartHookFunc, error) {
+	return PostStartHookName, func(hookContext genericapiserver.PostStartHookContext) error {
+		flowcontrolClientSet := flowcontrolclient.NewForConfigOrDie(hookContext.LoopbackClientConfig)
+		go wait.PollImmediateUntil(time.Minute, func() (bool, error) {
+			// ensure the pre-defined objects if the exempt flow-schema is present.
+			shouldEnsure, err := shouldEnsureAllPredefined(flowcontrolClientSet)
+			if err != nil {
+				klog.Errorf("failed getting exempt flow-schema, skipping creating default flowcontrol settings: %v", err)
+				return false, nil
+			}
+			if shouldEnsure {
+				if err := ensureAllPredefined(
+					flowcontrolClientSet,
+					flowcontrolbootstrap.PredefinedFlowSchemas(),
+					// Note: the "exempt" priority-level is supposed tobe the last item in the pre-defined
+					// list, so that a crash in the midst of the first kube-apiserver startup does not prevent
+					// the full initial set of objects from being created.
+					flowcontrolbootstrap.PredefinedPriorityLevelConfigurations(),
+				); err != nil {
+					klog.Errorf("failed creating default flowcontrol settings: %v", err)
+					return false, nil
+				}
+			}
+			return false, nil
+		}, hookContext.StopCh)
+		return nil
+	}, nil
+}
+
+// Returns false if there's a "exempt" priority-level existing in the cluster, otherwise returns a true
+// if the "exempt" priority-level is not found.
+func shouldEnsureAllPredefined(flowcontrolClientSet flowcontrolclient.FlowcontrolV1alpha1Interface) (bool, error) {
+	if _, err := flowcontrolClientSet.PriorityLevelConfigurations().Get(flowcontrol.PriorityLevelConfigurationNameExempt, metav1.GetOptions{}); err != nil {
+		if apierrors.IsNotFound(err) {
+			return true, nil
+		}
+		return false, err
+	}
+	return false, nil
+}
+
+// Creates the pre-defined priority-levels and then the flow-schemas the creation are supposed to be done in
+// 30s, otherwise an error will be returned.
+func ensureAllPredefined(flowcontrolClientSet flowcontrolclient.FlowcontrolV1alpha1Interface, flowSchemas []*flowcontrolv1alpha1.FlowSchema, priorityLevels []*flowcontrolv1alpha1.PriorityLevelConfiguration) error {
+	err := wait.PollImmediate(time.Second, 30*time.Second, func() (done bool, err error) {
```

Comment:
Why iterate here, when we also have an outer loop higher in the stack?

Why impose a 30-second time limit?

Reply:
the outer loop runs every one minute, the 30-sec limit makes sure the creation flow doesnt overlap.